### ISSUE TITLE
Loss And Training Experiments (Hinge / Smoothed / Sigma2)

### ISSUE THREAD

**Issue by bitsofbits** [GlobalFishingWatch/vessel-classification#181]
## Summary

Training was sometimes unstable. Reducing learning rate fixed that, but at the expense of 
much worse training outcomes.  I experimented with different objective functions and learning rate
setups to try to improve this problem.

## Experiments:

* I experimented with 3 new objective functions for classification:
  - Smoothing
  - Margin
  - Hinge

  Of these only Smoothing performed as well as baseline, and it's not clear that it was any better 
  than baseline. However it did make the training graph easier to interpret by reducing the 
  amount of bouncing in the output, so it is worth keeping a small amount of smoothing (0.1).

* I reduced the weight of the regression objectives (length, tonnage, etc). It immediately became
   apparent that a lot of the noise in the training and almost certainly much of the instability I
   was fighting came from the regression objectives. Probably because the training data is pretty
   noisy here. I experimented with with 0.01 and 0.1 weights and 0.1 seemed sufficient to fix the
   instability issue while still giving good performance for the regression outputs. [UPDATE], it
   appears that while 0.1 and 0.01 have similar performance 0.1 tends to favor larger classes
   over smaller classes (treating them more as noise presumably) and as a result small classes
   get under predicted.  There is a trade off between the type of weighting and and the amount
   of smoothing, but with sqrt weighting, 0.01 seems better in terms of correctly predicting
   less representative classes.

* I experimented with 1 "new" (really old) objective function for regression. This was motivated
  by the observation that the instability was primarily the result of the noisy training data for
  the regression values. So I tried the mean-absolute-error for the objective function.

*  I tried using sqrt weighting for the training samples again. The idea here is to split the difference
    between equal example weighting (where every example is seen the same number of times) and
    equal class weighting (where each class is seen the same number of times).  In this case, each
    example is weighted by sqrt(N_max / N_class), when N_class is the number of examples for a
    given class and N_max is the number of examples for the most numerous class.

    This significantly improved performance of the less numerous classes (on average) while also
    improving overall performance on the unweighted test set.

Overall sqrt-weighting and reducing the weight of regression objectives were solid improvements.
Light smoothing was also helpful for evaluating the run, while not hurting total performance and
MAE didn't have a large effect, but should in theory be more resistant to outliers, so we will leave in place.

**Comment by bitsofbits** [comment]
Tried several different modes including hinge loss. The best of these was smoothed loss (we only tried alpha = 0.1).
    ./deploy_cloudml.py --env dev --model prod.vessel_characterization_smoothed --job_name smoothed-1 --config_file deploy_characterization.yaml

The final run was actually stopped and started several times with tweaked learning rates, so should probably be rerun from scratch.

**Comment by bitsofbits** [comment]
gsutil cp gs://world-fishing-827-dev-ttl30d/data-production/classification/timothyhochberg/smoothed-1/models/prod.vessel_characterization_smoothed/train/model.ckpt-978411 .

        python -m classification.run_inference prod.vessel_characterization \
            --root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
            --inference_parallelism 128 \
            --feature_dimensions 14 \
            --inference_results_path=./smoother_test_978411.json.gz \
            --model_checkpoint_path ./model.ckpt-978411 \
            --metadata_file training_classes.csv \
            --fishing_ranges_file combined_fishing_ranges.csv \
            --dataset_split Test

**Comment by bitsofbits** [comment]
python compute_metrics.py     --inference-path ./smoother_test_978411.json.gz  --label-path classification/data/training_classes.csv                                 --dest-path vessel_char_smoothed.html --fishing-ranges classification/data/combined_fishing_ranges.csv --skip-localisation-metrics

**Comment by bitsofbits** [comment]
python -m classification.run_inference prod.vessel_characterization \
        --root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
        --inference_parallelism 128 \
        --feature_dimensions 14 \
        --inference_results_path=./smoother_test_978411.json.gz \
        --model_checkpoint_path ./model.ckpt-1029084 \
        --metadata_file training_classes.csv \
        --fishing_ranges_file combined_fishing_ranges.csv \
        --dataset_split Test

**Comment by bitsofbits** [comment]
gs://world-fishing-827-dev-ttl30d/data-production/classification/timothyhochberg/smoothed-5/models/prod.vessel_characterization_smoothed/train/model.ckpt-418069

python -m classification.run_inference prod.vessel_characterization_smoothed \
    --root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
    --inference_parallelism 128 \
    --feature_dimensions 14 \
    --inference_results_path=./smooth_slow_attr.json.gz \
    --model_checkpoint_path ./model.ckpt-418069 \
    --metadata_file training_classes.csv \
    --fishing_ranges_file combined_fishing_ranges.csv \
    --dataset_split Test

     python compute_metrics.py \
         --inference-path ./smooth_slow_attr.json.gz  \
         --label-path classification/data/training_classes.csv \
         --dest-path vessel_char_smoothed_slowattr.html \
         --fishing-ranges classification/data/combined_fishing_ranges.csv \
         --skip-localisation-metrics

**Comment by bitsofbits** [comment]
python -m classification.run_inference prod.vessel_characterization_smoothed \
--root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
--inference_parallelism 128 \
--feature_dimensions 14 \
--inference_results_path=./smooth_1_slow_attr.json.gz \
--model_checkpoint_path ./s1_model.ckpt-350590 \
--metadata_file training_classes.csv \
--fishing_ranges_file combined_fishing_ranges.csv \
--dataset_split Test

**Comment by bitsofbits** [comment]
```
python -m classification.run_inference prod.vessel_characterization_smoothed \
--root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
--inference_parallelism 128 \
--feature_dimensions 14 \
--inference_results_path=./smooth_1_slow_attr_b.json.gz \
--model_checkpoint_path ./s1_model.ckpt-431576 \
--metadata_file training_classes.csv \
--fishing_ranges_file combined_fishing_ranges.csv \
--dataset_split Test
```

**Comment by bitsofbits** [comment]
model.ckpt-368261-sqrt
python -m classification.run_inference prod.vessel_characterization_smoothed \
--root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
--inference_parallelism 128 \
--feature_dimensions 14 \
--inference_results_path=./smooth_1_slow_attr_sqrt_1.json.gz \
--model_checkpoint_path ./model.ckpt-368261-sqrt\
--metadata_file training_classes.csv \
--fishing_ranges_file combined_fishing_ranges.csv \
--dataset_split Test

**Comment by bitsofbits** [comment]
model-ckpt.662899-mae

python -m classification.run_inference prod.vessel_characterization_smoothed \
--root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
--inference_parallelism 128 \
--feature_dimensions 14 \
--inference_results_path=./smooth_1_slow_attr_sqrt_mae_1.json.gz \
--model_checkpoint_path ./model-ckpt.662899-mae \
--metadata_file training_classes.csv \
--fishing_ranges_file combined_fishing_ranges.csv \
--dataset_split Test
```

```
python compute_metrics.py \
--inference-path ./smooth_1_slow_attr_sqrt_mae_1.json.gz \
--label-path classification/data/training_classes.csv \
--dest-path vessel_char_smoothed_sqrt.html \
--fishing-ranges classification/data/combined_fishing_ranges.csv \
--skip-localisation-metrics 
```

**Comment by bitsofbits** [comment]
python -m classification.run_inference prod.vessel_characterization_smoothed \
--root_feature_path gs://machine-learning-dev-ttl-30d/classification/timothyhochberg/features-old-old-school/pipeline/output/features \
--inference_parallelism 128 \
--feature_dimensions 14 \
--inference_results_path=./smoothed_3.json.gz \
--model_checkpoint_path ./model.ckpt-801214-smoothed-3 \
--metadata_file training_classes.csv \
--fishing_ranges_file combined_fishing_ranges.csv \
--dataset_split Test

python compute_metrics.py \
--inference-path ./smoothed_3.json.gz \
--label-path classification/data/training_classes.csv \
--dest-path smoothed_3.json.html \
--fishing-ranges classification/data/combined_fishing_ranges.csv \
--skip-localisation-metrics \
--dump-labels-to smoothed_labels \
--dump-fine-labels-to smoothed_fine_labels \
--dump-attributes-to smoothed_attributes
